README
================================================================================
                                 Bookmark Tools
================================================================================

    Bookmark tools is a collection of useful shell commands and Python scripts
    that aim to speed up navigation and organization in day to day work. 

================================================================================

    Commands

    bm                  Directory bookmark utility
    cdbm                Navigate to bookmark
    lsbm                List files in bookmark
    cpbm                Copy to bookmark    
    lnbm                Place link in bookmark
    mvbm                Move files to bookmark
    mkbm                Make new directory and bookmark
    blog                A minimalist and extendable publishing tool 
    fill-template       A simple template language
    script2gif          Makes an animated gif out of shell commands and stdout
    bm-match            Weighted file sorting 
    bmsuggest           Suggest bookmarks for files
    bmsuggest-move      Move files to suggested bookmark
    bmsuggest-series    Move files to suggested bookmark and add a serie nr.
    bm-add-series       Adds a serie nr. from the file name to a bookmark


================================================================================

    F.A.Q.


    *** Where can I find out more about using the tools?

    The man pages should provide you with information on the options and
    arguments each of the tools accept. You can also browse these documents 
    on-line, which might be more pleasent: http://www.bookmark-tools.com/
    Other tutorials and tips should also become available on that website.


    *** Can I use this with zsh?

    Yes, all the tools and shell functions are completely functional and miss
    only one thing that: tab completion. Adding this has currently a very low
    priority for me, since I mostly use bash, but I'll gladly accept your
    patches ;)


    *** Is there a source repository I could follow?

    Certainly, there is a git repository which you can clone with the following
    command: `git clone git://www.bookmark-tools.com/btools.git btools`
    It's also browsable on-line. See http://www.bookmark-tools.com/git/


================================================================================

    Trouble-shooting

    *** bash: cdbm: command not found

    First of all; make sure you have restarted your shell since installation.
    Most of the commands will be installed in /usr/bin. Others are implemented
    via shell functions and need to be sourced before they can be used. 
    This is handled automatically by the setup script, which adds a source
    directive to /etc/bash.bashrc which should make the functions available in
    every login shell. This does mean you have to restart to make full use
    of the tools.

    When the cdbm command still can't be found, it's safe to assume that
    /etc/bash.bashrc is not loaded by default on your platform. To work around
    this issue you can source /usr/share/bm/bm.bash in your own ~/.bashrc file
    (or equivalent) by adding the line:

        source /usr/share/bm/bm.bash


================================================================================

    Resources

    [1] http://www.bookmark-tools.com/
        The official website
    [2] http://groups.google.com/group/btools
        A Google group for discussions/bug reports/commit logs.

================================================================================

    PLUGS!

    If you like music and you like Python, why not check out mingus? It's a
    general purpose music library with a small group of contributors that will 
    make your life so much nicer. You're gonna love your notes. 
    http://mingus.googlecode.com/

================================================================================

    Bookmark Tools, navigation and organization tools
    Copyright (C) 2009, Bart Spaans <bart@bookmark-tools.com>

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


================================================================================


CHANGELOG
commit c99ceb7adcd80529ae3fe15e0082b6b31bd60dbc
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 05:34:05 2009 +0200

    Update to incorperate script2gif tool

commit feeddacf12791bb57a78e0dfbf61f0e8d07074c9
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 05:29:40 2009 +0200

    Updated docs

commit 9a6b97c7919cc91abed8a7d72de1aefe5f42fc7b
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 05:29:11 2009 +0200

    Updated mkbm docs and common interface

commit 03a64ce21751eb38fe332da84fbd5515e19c43cf
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 05:17:42 2009 +0200

    Some updates in the script2gif tool

commit 783ca86fa09b224ef1d6b48c171e3e5405bf3974
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 00:52:31 2009 +0200

    Updated docs again..

commit a1bf210ed80c12463a1f05b5692510b6ca40fde7
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 00:50:26 2009 +0200

    Updated documentation.

commit b9fdbb85e56c62e1d99fb259c44b53bfbb1d9482
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Sep 2 00:35:32 2009 +0200

    Added script2gif tool
    
    Produces an animated gif from shell commands and their output.

commit 7a0b86c22bc9fe6274fe843efd3c0e31bfdebd54
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Tue Sep 1 23:40:57 2009 +0200

    Added some fixes to README and setup description

commit 1a5b02a28306c6fd226ba850a6730d92206789c0
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 05:45:55 2009 +0200

    Fixed bugs in setup script

commit 5363363e370498a9ee65aff1868fcdb9c22dcee5
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 04:42:53 2009 +0200

    Updated setup script for PyPI submission.
    
    Going to try to submit version 0.999 to PyPI so it can be used with
    easy_install and such. Up next: working towards 1.0rc1.

commit 8ed59a666594d4727b59b0cea5a26dd6ae58876a
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 04:36:17 2009 +0200

    Added LICENSE and CHANGELOG

commit f003c1ca632c0c6161eec2da441c828f55d1d538
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 04:34:50 2009 +0200

    Updated README and decided on the license.

commit af1fc67b66ea600ada597d82164e8fa376e92c11
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 04:34:40 2009 +0200

    Added completion to the bm command itself

commit 6413f490bfd8d3b08aaac4f6e00599a42a38b480
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 03:27:59 2009 +0200

    Updated manifest.in and todo

commit 0c9d2822270792b2ceb1d4e61b23e5e6242b751e
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 03:26:08 2009 +0200

    Removed some bashisms from bm.bash
    
    I'm testing out the functions with zsh and it wouldn't work because of
    the way I tested for equality with ==. I now replaced those tests with
    appropriate -eq and =.

commit dc4a10bb96bdbe83c9fb468f11ebe01ff664791e
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Wed Aug 26 03:25:27 2009 +0200

    Added INSTALL and README files

commit bc1eea551bed53017817e45564c27764bc3580aa
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 15:20:49 2009 +0200

    Added more extended shell commands.
    
    Added lsbm, lnbm and scpbm to the shell functions.

commit f8e8cb2c3e0f100720096047187705725a171c69
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 15:06:34 2009 +0200

    Added bash completion to blog tool

commit 06fbc82b350f225a91c414fda3f96b5145ffaee2
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 12:16:03 2009 +0200

    Fixed empty index bug in bm

commit a10b42e6d1da42e0f9cc91bd448598a736dcae77
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 11:52:35 2009 +0200

    Setup script: sourcing data/bm.bash in /etc/bash.bashrc
    
    This means that the cdbm command will now automatically be available
    after installation.

commit a495fe5da79f934883e44295150fbdf277d93c95
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 10:35:41 2009 +0200

    Fixed 'blog rm OBJECTID' bug

commit fa4786225262395eff538be3ae7fbe4b0ad701ca
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Mon Aug 24 09:07:12 2009 +0200

    Made verbose log more verbose

commit 92635ee6f0d211d2289abd1d3863df896edf1a13
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 14:44:28 2009 +0200

    Updated blog log docs

commit f57c980f78d2deee405af8fba3b2e30174a78a5f
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 14:30:30 2009 +0200

    Finishing touches in blog tool

commit 9fda6bf284c5c2296de75b4b265297c527d7822d
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 13:09:44 2009 +0200

    Blog is fully functional

commit 10b5aa885bc9279b43fade28fb2251dbd53fc345
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 12:56:39 2009 +0200

    Generated new documentation for fill-template

commit 2716c76cc297accb61f6cb8436e01e6c1d28dd05
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 12:55:33 2009 +0200

    Added new fill-template tool

commit 39b8329eee7fc020751b4aa5ecb9734adb463c2c
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Fri Aug 21 11:21:44 2009 +0200

    Added publish-new and publish-edited hooks

commit 4ae679aae9ca13f0fde9e1c00c185ae8513bd6ac
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Tue Aug 18 07:25:47 2009 +0200

    Updated build_man script to use module's long_description if present.

commit 5188e8c12477268a7adb1ee1dddea490cc18f7a7
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Tue Aug 18 06:59:43 2009 +0200

    Started on extra command line options in blog

commit 9b51607b3e670f0b287b9f1773a2833f88795012
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Tue Aug 18 06:12:25 2009 +0200

    Updated documentation.

commit 5a1744ca515318034c51ce259ccf5280677d92e3
Author: Bart Spaans <onderstekop@gmail.com>
Date:   Tue Aug 18 05:47:34 2009 +0200


btools/matching/RuleParsers.py
from Rule import Rule
import btools.common as common

import os
import re
import subprocess
import operator as op
import time

class Parser:

    match_token = [""]
    help = ""

    def __init__(self):
        self.oper = { '==': op.eq, '<=': op.le, "<": op.lt, 
                     '>': op.gt, '>=': op.ge, '!=': op.ne,
                     'eq': op.eq, 'le': op.le, 'lt': op.lt,
                     'gt': op.gt, 'ge': op.ge, 'ne': op.ne}
        self.init()

    def init(self):
        pass

    def get_Rule(self, slist, bookmark, weight):
        def logged_match(m, file, variables):
            common.debug("Testing %s against %s rule (%s)" % (file, self.match_token[0], " ".join(slist)))
            return m(file, variables)

        r = Rule()
        match = self.get_match_function(slist, {"bookmark": bookmark, "weight": weight})
        if match is not None:
            r.bookmark = bookmark
            r.weight = weight
            r.text = " ".join(slist)
            r.match_func = lambda file, variables: logged_match(match, file, variables)
            r.match_token = self.match_token[0]
            return r

    def get_match_function(self, slist, target):
        pass

    def __str__(self):
        return "%s parser" % (self.match_token[0].capitalize())

    def __repr__(self):
        return self.__str__()



class ExtensionParser(Parser):

    match_token = ["extension", "extensions", "ext"]

    def get_match_function(self, slist, target):
        ext = []
        for x in slist:
            x = x.strip()
            if x != "":
                if x[-1] == ",":
                    x = x[:-1]
                if x[0] == ".":
                    x = x[1:]
                ext.append(x.lower())

        def match(file, variables):
            extension = file.split(os.path.extsep)
            if os.path.isfile(file) and len(extension) > 1:
                extension = extension[-1]
                return extension.lower() in ext
            return False
        return match

class SizeParser(Parser):

    match_token = ["filesize", "size"]

    def init(self):
        rfilesize = "(==|<=|<|>|>=|!=|eq|le|lt|gt|ge|ne){1}\ *([0-9]*[\.[0-9]*]?)\ *(b|k|M|G|P|T|P)+?b?\ *$"
        self.re_filesize = re.compile(rfilesize, re.IGNORECASE)


    def get_match_function(self, slist, target):
        sizedict = { 'b': 1, 'k': 1 << 10, 'm': 1 << 20, 'g': 1 << 30,
                    't': 1 << 40, 'p': 1 << 50 }

        s = " ".join(slist)
        r = self.re_filesize.search(s)
        if r is None: return None

        g = r.groups()
        size = int(sizedict[g[2].lower()] * float(g[1]))
        cmp = self.oper[g[0].lower()]

        def match(file, variables):
            if os.path.isfile(file):
                return cmp(os.path.getsize(file), size)
            return False

        return match

class TimeParser(Parser):

    match_token = ["time"]

    def init(self):
        rtime = "(==|<=|<|>|>=|!=|lt|le|eq|ge|gt|ne){1}\ *([0-9]*[\.[0-9]*]?)\ *(y|mo|w|d|m|h|s)+"
        self.re_time = re.compile(rtime, re.IGNORECASE)

    def get_match_function(self, slist, target):
        timedict = { "s": 1, "m": 60, "h": 3600, "d": 86400, 
                     "w": 604800, "mo": 2592000, "y": 31536000 }

        s = " ".join(slist)
        r = self.re_time.search(s)
        if r is None: return None

        g = r.groups()
        pivot = float(g[1]) * timedict[g[2]]
        cmp = self.oper[g[0]]

        def match(file, variables):
            if os.path.isfile(file):
                t = os.path.getmtime(file)
                tnow = time.time()
                return cmp(t, tnow - pivot)
        return match

# Opens a new process
class ExpressionParser(Parser):

    match_token = ["expression", "expr"]

    def get_match_function(self, slist, target):
        s = " ".join(slist)
        def match(file, variables):
            c = common.replace_variables(s.replace("%file%", pipes.quote(file)), variables)
            try: return subprocess.Popen(c, shell = True).wait() == 0
            except: pass
            return False
        return match


#
#   Regular Expression Parsers
#

class RegexParser(Parser):
    
    match_token = ["regex"]

    def get_regex_flags(self):
        return 0

    def get_match_function(self, slist, target):
        regex = re.compile(" ".join(slist), self.get_regex_flags())
        def match(file, variables):
            f = file.split(os.path.sep)
            if f == []:
                return False
            file = f[-1]
            if file == "" and len(f) > 1:
                file = f[-2]
            if regex.search(file):
                return True 
            return False
        return match


class iRegexParser(RegexParser):

    match_token = ["iregex"]

    def get_regex_flags(self):
        return re.IGNORECASE


class SelfParser(iRegexParser):

    match_token = ["self", "itself"]

    def get_match_function(self, slist, target):
        return RegexParser.get_match_function(self, [target["bookmark"]], target)


class SentenceParser(iRegexParser):

    match_token = ["sentence", "words"]

    def get_match_function(self, slist, target):
        return RegexParser.get_match_function(self, [".*".join(slist)], target)





